### DISCUSSION TITLE
question on off-diagonal (nonlocal?) jacobian for nodal kernels

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#23196] [Q&A General] [answered]
With Nodal Kernels, is it possible to set the Jacobian contribution of a nodal value with respect to a different node **_for the same variable_**? My understanding of the way off-diagonal Jacobians usually work in MOOSE is you define the contribution of one variable with respect to another variable **_at a given node/qp_**. What I'm trying to do is set the Jacobian for $\frac{\partial u_i}{\partial u_j}$, say, for an $N$-node system with one nonlinear variable, and make a dense $N$ x $N$ Jacobian.

Is this what the `computeNonlocalOffDiagJacobian` member function is for? If so, are there any examples/documentation on how to use this?

**Comment by GiudGiud** [comment, marked as answer]
Hello

NonLocalIntegratedBC seems to be an example doing that.
I wonder, could you just AD to let it process the off diagonals?

Guillaume

**Reply by jessecarterMOOSE** [reply]
That seems close but unfortunately I don't understand how the Assembly stuff works and how the global residual and Jacobian get numbered/stored. Is that kind of info available to read up on? What changes would need to be made for a NodalKernel?

Not sure if AD will work in my case because my contribution to the residual/Jacobian includes a Postprocessor value. But if I had the nonlocal variable values then I wouldn't need the Postprocessor in the first place.

**Reply by GiudGiud** [reply]
There's the doxygen for assembly.
I dont think we have much on the numbering, the main documentation https://mooseframework.inl.gov/moose/source/systems/NonlinearSystem.html has a lot but not that

Looking at the doxygen for the base nodal kernel, you also have that routine that you can implement if there s a nonlocal off-diagonal contribution to the jacobian
https://mooseframework.inl.gov/docs/doxygen/moose/classResidualObject.html#a706e1d22b9aad2f08bed44606afced07

**Reply by jessecarterMOOSE** [reply]
Thanks @GiudGiud . Though in my case, I guess I want computeNonlocalJacobian, right? The OffDiag bit refers to derivative with respect to a different variable? I'm looking at derivative with respect to same variable but different DoF.

**Reply by GiudGiud** [reply]
It's still going to be off-diagonal?
The diagonal is only for the local element and local variable

**Reply by jessecarterMOOSE** [reply]
I guess I'm confused about "on" and "off" diagonals because I can't see how the system is constructed. What's the reason for having computeNonlocalJacobian and computeNonlocalOffDiagJacobian?

**Reply by GiudGiud** [reply]
not sure. You may be right about the distinction. 

@lindsayad will know.

**Reply by lindsayad** [reply]
OffDiag generally denotes a different variable and on-diagonal refers to same variable, so these API names are in the block matrix sense.

**Reply by lindsayad** [reply]
@dschwen and @SudiptaBiswas are the nonlocal authors, so maybe one of them could comment on those APIs
